pkg/resource/deploy/source_eval.go

```diff
@@ -566,7 +570,11 @@ func (rm *resmon) RegisterResource(ctx context.Context,
 	req *pulumirpc.RegisterResourceRequest) (*pulumirpc.RegisterResourceResponse, error) {
 
 	// Communicate the type, name, and object information to the iterator that is awaiting us.
-	t := tokens.Type(req.GetType())
+	t, err := tokens.ParseTypeToken(req.GetType())
```

Comment:
IMO we should only enforce this for custom resources:
- We only really require the type in order to look up a resource's default provider, and component resources do not use providers
- It is not at all clear when authoring a component what the module portion of the component's type should be

(not to mention that this is a breaking change for components)

Reply:
Note that we do have code in the service that parses these things and expects a certain format. We should just make sure whatever we do here continues to operate correctly there.

Reply:
> Note that we do have code in the service that parses these things and expects a certain format

The service is already broken w.r.t. component resources, then, and we should fix it.

Reply:
In my mind it's dangerously unsound to allow things of type `tokens.Type` to exist in our system that do not fulfill the refinements that `tokens.Type` claims to have.

> We only really require the type in order to look up a resource's default provider, and component resources do not use providers

That is true, but this argument also applies to e.g. passing a null pointer to a function that requires a non-nil pointer. Sure, it's a refinement violation, but it's only a problem if we use it and we only use the pointer in X specific case.

I fear that we will have endless issues of this sort unless we attempt to reclaim at least some of our invariants.

> It is not at all clear when authoring a component what the module portion of the component's type should be

This is true, but is this actually a problem? Anyone can come up with whatever they want here. It's also natural to reach for the two-colon syntax for type names because all other resources in our system use that convention. Component resources are still resources and I believe that it makes sense to apply a consistent type naming convention across them.

> (not to mention that this is a breaking change for components)

The only breaking change here is that we won't allow "primitive" types (i.e. single word identifiers) as a type. If we're not OK with this, it's an easy check in `ParseTypeToken`. It is arguably a bug that we allowed it at all.

Reply:
> It's also natural to reach for the two-colon syntax for type names because all other resources in our system use that convention.

99% of the time a user will never type anything using this syntax when authoring a program. Authoring component resources is essentially the only time that this arises.

> The only breaking change here is that we won't allow "primitive" types (i.e. single word identifiers) as a type.

Is this true? I think we were previously allowing anything besides the empty string as the type of a component resource... Regardless, I am more comfortable with your suggestion to expand the allowed type format for components to include single-word identifiers, as that captures what I'd guess is the majority of the cases.

Note that breaking this might be pretty bad, as I _think_ that fixing it will end up causing URNs for component resources _and their children_ to change, which ends up causing all of these things to be deleted and recreated.

Reply:
I'm not sure we want single word identifiers. For better or for worse, we use these things all over the place -- including the service -- when displaying information, and we generally interpret the package in meaningful ways. The module is usually superfluous, but we definitely treat it specially. I think we want a deeper design discussion if we're going to change this and I would personally vote for going with Sean's change as originally conceived. At least we get to a point where our system is working with proper invariants, and then we can decide to change it in a methodical way. Just my $0.02.

Reply:
> Is this true? I think we were previously allowing anything besides the empty string as the type of a component resource...

The 0.15 CLI crashes when presented with a one-colon *component* resource type: https://github.com/pulumi/pulumi/issues/1880. Customers have already been broken - I don't think it makes sense to fear a breaking change when the damage has already been done.

https://github.com/pulumi/pulumi/issues/1880 also illustrates why I think it's dangerous to allow invalid `tokens.Type` to flow through our system - it is *incredibly easy* to slip up and forget to check if a resource is custom before attempting to pull packages out of a type. That particular bug is a bug in `deploy/plan.go` where one of our conditionals checks `IsProviderType` before `res.Custom`.